src/exchange/moderators/summarizer.py

```diff
@@ -1,135 +1,49 @@
-            if throwaway_exchange.messages[0].tool_result:
-                m, c = pop_checkpoint(throwaway_exchange, exclude_last=1, return_messages=True)
-                messages_to_summarize.extend(m)
-                exclude_last = True
-
-            if messages_to_summarize[-1].role == "assistant" and (not messages_to_summarize[-1].tool_use):
-                messages_to_summarize.append(Message.user("Summarize our the above conversation"))
-
-            summarizer_exchange = exchange.replace(
-                system=Message.load("summarizer.jinja").text,
-                moderator=ContextTruncate(),
-                model=self.model,
-                messages=messages_to_summarize,
-                checkpoints=[],
-            )
-
-            summary = summarizer_exchange.reply()
-            summary_checkpoint = summarizer_exchange.checkpoints[-1]
-
-            # pop out all the messages up
-            for _ in range(removed_checkpoints):
-                pop_checkpoint(exchange)
-
-            # if first message is a tool_result, it means there's no matching pair. remove it
-            if exclude_last:
-                pop_checkpoint(exchange, exclude_last=1)
-
-            # insert summary as first message/checkpoint
-            if exchange.messages[0].role == "assistant":
-                summary_message = Message.user(summary.text)
-            else:
-                summary_message = Message.assistant(summary.text)
-
-            exchange.messages.insert(0, summary_message)
-
-            # insert a new checkpoint with the summary message
-            exchange.checkpoints.insert(
-                0,
-                Checkpoint(
-                    start_index=0,
-                    end_index=1,
-                    token_count=summary_checkpoint.latest_generated_tokens + self.system_prompt_token_count,
-                    latest_generated_tokens=summary_checkpoint.latest_generated_tokens,
-                ),
-            )
 
-            # update checkpoint indices
-            for cp in exchange.checkpoints[1:]:
-                cp.start_index = cp.start_index + 1
-                cp.end_index = cp.end_index + 1
+        self._update_system_prompt_token_count(exchange)
+
+        if exchange.checkpoint_data.total_token_count < self.max_tokens:
+            return
+
+        messages_to_summarize = self._get_messages_to_remove(exchange)
+        num_messages_to_remove = len(messages_to_summarize)
+
+        # the llm will throw an error if the last message isn't a user message
+        if messages_to_summarize[-1].role == "assistant" and (not messages_to_summarize[-1].tool_use):
+            messages_to_summarize.append(Message.user("Summarize our the above conversation"))
+
+        summarizer_exchange = exchange.replace(
+            system=Message.load("summarizer.jinja").text,
+            moderator=PassiveModerator(),  # note-pr-reviewer: this was previously ContextTruncate
+            model=self.model,
```

Comment:
note: this used to be a truncate moderator, but we now are using a passive one. why? because we know the message here will always be under the token limit